core/local/layers/linux.js

```diff
@@ -0,0 +1,180 @@
+//
+// On Linux, the API to watch the file system (inotify) is not recursive. It
+// means that we have to add a watcher when we a new directory is added (and to
+// remove a watcher when a watched directory is removed).
+//
+// Ignoring some files/folders could have been done in a separated layer, but
+// it is more efficient to do here, as we can avoid to setup inotify watchers
+// for ignored folders.
+//
+// Even if inotify has a IN_ISDIR hint, atom/watcher does not report it. So, we
+// have to call stat on the path to know if it's a file or a directory for add
+// and update events.
+module.exports = class LinuxSource {
+  /*::
+  syncPath: string
+  next: Layer
+  running: boolean
+  watchers: Map<string, *>
+  */
+  constructor (syncPath /*: string */, next /*: Layer */) {
+    this.syncPath = syncPath
+    this.next = next
+    this.running = false
+    this.watchers = new Map()
+    autoBind(this)
+  }
+
+  async start () {
+    this.running = true
+    await this.watch('.')
+    this.next.initial()
+  }
+
+  async watch (relativePath /*: string */) {
+    try {
+      const fullPath = path.join(this.syncPath, relativePath)
+      const w = await watcher.watchPath(fullPath, {}, this.process)
+      if (!this.running) {
+        w.dispose()
+        return
+      }
+      this.watchers.set(relativePath, w)
+      const batch /*: LayerEvent[] */ = []
+      for (const entry of await fse.readdir(fullPath)) {
+        try {
+          const abspath = path.join(this.syncPath, relativePath, entry)
+          batch.push(await this.buildAddEvent(abspath))
+        } catch (err) {
+          // TODO error handling
+        }
+      }
+      // TODO ignore
+      if (batch.length === 0) {
+        return
+      }
+      this.next.process(batch)
+      for (const event of batch) {
+        if (event.doc.docType === 'folder') {
+          await this.watch(event.doc.path)
+        }
+      }
+    } catch (err) {
+      // The directory may been removed since we wanted to watch it
+    }
+  }
+
+  async process (events /*: AtomWatcherEvent[] */) {
+    // TODO logger
+    // TODO ignore
+    const batch /*: LayerEvent[] */ = []
+    for (const event of events) {
+      switch (event.action) {
+        case 'created':
+          const eAdd = await this.buildAddEvent(event.path)
+          batch.push(eAdd)
+          if (eAdd.doc.docType === 'folder') {
+            this.watch(eAdd.doc.path)
+          }
+          break
+        case 'modified':
+          batch.push(await this.buildUpdateEvent(event.path))
+          break
+        case 'deleted':
+          const eDel = await this.buildRemoveEvent(event.path, event.kind)
+          batch.push(eDel)
+          const w = this.watchers.get(eDel.doc.path)
+          if (w) {
+            w.dispose()
+            this.watchers.delete(eDel.doc.path)
+          }
+          break
+        case 'renamed':
+          batch.push(await this.buildMoveEvent(event.path, event.oldPath))
+          break
```

Comment:
I assume existing watchers automatically keep watching the new destination?

Reply:
Yes, inotify watchers are attached to a folder and move with it.